### PR TITLE
Common Subgraph Elimination

### PR DESCRIPTION
**Pull request by mzient**
## Category:
**New feature** (*non-breaking change which adds functionality*)

## Description:
This change introduces a graph-level optimization that removes common subgraphs.
The algorithm goes over graph nodes in topological order and computes a key from the operator's schema name, arguments and (renamed) inputs. If the operator is found, it's replaced with the previously found one and the output names are added to the renaming map. When subsequent operators are processed, the inputs of the replaced operator are renamed accordingly.
Example:
```
A = foo(arg=42)  # new
B = foo(arg=42)  # replaced with A
C = bar(A) # new
D = bar(B) # replaced with bar(A)
return C + D  # replaced with C + C
```
This optimization is particularly useful with constants, the use of `.shape` function and subscripts.
In the following example:
```
horz_center = (bbox[0] + bbox[2]) / 2
width = (bbox[2]  - bbox[0])
```
The `bbox[0]` and `bbox[2]` will be reused.

### Tests:
Some Python tests needed to be adjusted.

- [X] Existing tests apply
- [X] New tests added
  - [X] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [ ] Existing documentation applies
- [x] Documentation updated
  - [ ] Docstring
  - [x] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [X] N/A

**REQ IDs**: N/A

**JIRA TASK**: DALI-4145

### CODE REVIEW

File: dali/pipeline/operator/argument.cc [hunk 1]
```diff
@@ -41,7 +41,8 @@ inline std::shared_ptr<Argument> DeserializeProtobufVectorImpl(const DaliProtoPr
   auto args = arg.extra_args();
   std::vector<T> ret_val;
   for (auto& a : args) {
-    const T& elem = DeserializeProtobuf(a)->Get<T>();
```
Comment (mzient) [inline dali/pipeline/operator/argument.cc:44]: After changing `Argument::Get` to returning a reference, this became a bug: returning a reference to a field of the result of `DeserializeProtobuf`.

File: dali/test/python/test_pipeline.py [hunk 2]
```diff
@@ -574,18 +574,17 @@ def define_graph(self):
...
         pipe.deserialize_and_build(s)
         pipe_out = pipe.run()
         pipe_out_cpu = pipe_out[0].as_cpu()
-        img_chw_test = pipe_out_cpu.at(n)
         if i == 0:
-            img_chw = img_chw_test
-        assert np.sum(np.abs(img_chw - img_chw_test)) == 0
+            ref = pipe_out_cpu
+        else:
+            check_batch(pipe_out_cpu, ref)
```
Comment (mzient) [inline dali/test/python/test_pipeline.py:587]: This test checked too little and took too long - now it's more accurate and much faster.

File: dali/pipeline/graph/cse.cc [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+  }
+
+  return op.SerializeAsString();
+}
+
+/** The context for Common Subgraph Elimination */
+class CSE {
+ public:
+  void Run(OpGraph &graph) {
+    for (auto &node : graph.OpNodes())
```
Comment (banasraf) [inline dali/pipeline/graph/cse.cc:74]: OpNodes returns the nodes in topological order?

Comment (mzient) [reply dali/pipeline/graph/cse.cc:74]: Yes, the graph is sorted by the Builder.

File: dali/pipeline/operator/op_schema.h [hunk 4]
```diff
@@ -89,11 +89,12 @@ struct ArgumentDef {
...
-  bool tensor     = false;
-  bool per_frame  = false;
-  bool internal   = false;
-  bool hidden     = false;
+  bool required   = false;  //< The argument must be set.
+  bool tensor     = false;  //< The argument can be provided as a TensorList
+  bool per_frame  = false;  //< The (tensor) argument can be expanded to multiple frames
+  bool internal   = false;  //< The argument cannot be set by the user in Python
+  bool hidden     = false;  //< The argument doesn't appear in the documentation
+  bool ignore_cmp = false;  //< Two operators can be considered equal if this argument differs
```
Comment (szkarpinski) [inline dali/pipeline/operator/op_schema.h:97]: Nitpick
```suggestion
  bool ignore_cmp = false;  //< Two operators can be considered equal even if this argument differs
```

File: dali/pipeline/operator/op_schema.cc [hunk 5]
```diff
@@ -76,6 +97,10 @@ OpSchema::OpSchema(DefaultSchemaTag) : name_(""), default_(true) {
   AddInternalArg("default_cuda_stream_priority", "Default cuda stream priority", 0);  // deprecated
   AddInternalArg("checkpointing", "Setting to `true` enables checkpointing", false);
 
+  AddInternalArg("preserve_name", R"(When true, the operator cannot be renamed.
+This disables merging this operator with another one with a different name.)",
+                 false);
```
Comment (szkarpinski) [inline dali/pipeline/operator/op_schema.cc:102]: Shouldn't this be the case **always** when an operator was explicitly named by the user?

I understand that having two equal ops (Readers?) but with different names is not a clever way to use DALI, but it is possible for the user to do that. Let's say that for whatever reason I have two readers doing the same work, but named differently ("Reader1", "Reader2"). I wrapped the pipeline into something (an iterator?) that relies on a presence of "Reader2" in the pipeline. This is not a good code, but it works. And after CSE it doesn't

Comment (mzient) [reply dali/pipeline/operator/op_schema.cc:102]: That's exactly how it works. If you specify a name manually, then `preserve_name` is set to `True`.

Comment (szkarpinski) [reply dali/pipeline/operator/op_schema.cc:102]: Ah, I see:
```
self._spec.AddArg("preserve_name", not self._autoname)
```
Thank you!

File: dali/test/python/test_pipeline.py [hunk 6]
```diff
@@ -2333,3 +2332,61 @@ def def_ref():
         (ref,) = ref_pipe.run()
         check_batch(cpu, ref, bs, 0, 0, "HWC")
         check_batch(gpu, ref, bs, 0, 0, "HWC")
+
+
+def test_cse():
```
Comment (szkarpinski) [inline dali/test/python/test_pipeline.py:2337]: How about adding a test with External Source (to make sure it's not eliminated) and manually named operators?

Comment (mzient) [reply dali/test/python/test_pipeline.py:2337]: Manually named operator: see line 2343.
ES - we already have tests with multiple external sources and they did fail before adding the fix, so I'd call it covered.

**Review by banasraf — APPROVED** [review]

**Review by szkarpinski — APPROVED** [review]

**Pull request merged**